### ISSUE TITLE
Middelman preview and reload very slow

### ISSUE THREAD

**Issue by iwarner** [middleman/middleman#1860]
Guys

I know this has been stated a few times, but as my project grows I am running into a constant battle of the reload being extremely poor

I am still using sprockets but recently added the sassc not much happened I think

I do use a few sprockets appends
sprockets.append_path File.join( root, "bower_components" )
sprockets.append_path File.join( root, "source/atom" )
sprockets.append_path File.join( root, "source/molecule" )
sprockets.append_path File.join( root, "source/organism" )
sprockets.append_path File.join( root, "source/template" )
sprockets.append_path File.join( root, "source/page" )

Which could be an issue?

Anyway if I can ask if you can benchmark my project :
https://github.com/iwarner/CodeBlender-Middleman 

It could help with your development also?

This is the Gist of my middlemen preview server verbose
https://gist.github.com/iwarner/2783130afe18d2a876f6

Main pause is at line 1170 
Just before it starts the server - [2016-03-24 12:14:26] INFO  WEBrick 1.3.1

Gems installed
middleman (4.1.4)
middleman-cli (4.1.4)
middleman-core (4.1.4)
middleman-favicon-maker (4.0.3)
middleman-livereload (3.4.6)
middleman-minify-html (3.4.1)
middleman-syntax (2.1.0)

I always use : "bundle exec middleman server" to initiate

Many thanks

**Comment by tdreyno** [comment]
Last time I benchmarked this, it was mostly coming from `sprockets`. Ping @stevenosloan for visibility.

**Comment by stevenosloan** [comment]
Hmmm, yeah it could be -- it doesn't look like there's a terribly large number of files that would be handled by `SprocketsResource`'s so I wouldn't think that timing would be too long (2-3 seconds as a guess). I'm out on some terrible internet atm, but will hopefully have a better chance to dig in tomorrow.

**Comment by iwarner** [comment]
Guys
Many thanks for checking this out - Just so you know its 

40 secs for server load 

    == The Middleman is loading
   == Preferring use of LibSass
   == Sprockets will render css with SassC
   == Locales: en (Default en)
   == Blog Sources: /docs/:title.html (:prefix + :sources)
   == View your site at "http://Ians-Air.home:4567", "http://192.168.1.91:4567"
   == Inspect your site configuration at "http://Ians-Air.home:4567/__middleman", "http://192.168.1.91:4567/__middleman"

After one file change 

DomContetLoaded = 12.91s
Load = 12.91s

My JS is 1.7mb as I am testing a lot of libraries.

Refresh with no file change

Load = 3s

**Comment by tdreyno** [comment]
I'll investigate. Blog is really slow right now, will let you know.

**Comment by tdreyno** [comment]
Changed a nested `all_files.none?` to `!Set.new(all_files).include?` and saved 30 seconds. Hurray?

Still showing 5s spent in Sprockets, via `middleman server --instrument`

**Comment by iwarner** [comment]
Hiya is that on the server start - the refresh is really what is killing me?

Takes 12 seconds to refresh the page on a change...

Let me try your fix anyway and report back

**Comment by iwarner** [comment]
Yay 15 second server launch.. nice one

Unfortunately refreshing the page is still upwards of 11 seconds, so I think this needs re-opening to see whats going on when a file is changed.

**Comment by stevenosloan** [comment]
The refresh issue I *think* may be sprockets related. Looking at being a bit smarter about invalidating it's cache. Are you seeing that 15s on *every* refresh or just after a reload/file change?

**Comment by iwarner** [comment]
Hiya, 

After one file change
Load = 12.91s

Refresh with no file change
Load = 3s

**Comment by tdreyno** [comment]
What do you mean "refresh the page"? Like, literally hitting the browser refresh button?

If the sprockets manipulator is ~5s, a file change might end up being that long.

@iwarner Can you run your server with the `--instrument` flag and report back the results.

**Comment by iwarner** [comment]
Sure - yeah literally cmd+r = 12 seconds to load.

I am running 4.1.6 now

--instrument as follows : 

== The Middleman is loading
== Preferring use of LibSass
== Sprockets will render css with SassC
== Instrument (sitemap.manipulator): 31.099999999999998ms
{:name=>:sitemap_ondisk}
== Instrument (sitemap.manipulator): 167.987ms
{:name=>:front_matter}
== Instrument (sitemap.manipulator): 6987.7029999999995ms
{:name=>:sprockets}
== Instrument (sitemap.update): 7240.716ms
{:reasons=>[:touched_file, :registered_new_manipulator_front_matter, :registered_new_manipulator_sitemap_ondisk, :registered_new_manipulator_sitemap_import, :registered_new_manipulator_sitemap_endpoint, :registered_new_manipulator_sitemap_proxies, :registered_new_manipulator_sitemap_redirects, :registered_new_manipulator_sitemap_move_files, :registered_new_manipulator_sitemap_ignore, :registered_new_manipulator_routing, :registered_new_manipulator_collections, :registered_new_manipulator_sprockets]}
== Locales: en (Default en)
== Blog Sources: /docs/:title.html (:prefix + :sources)
== Instrument (sitemap.manipulator): 48.44ms
{:name=>:sitemap_ondisk}
== Instrument (sitemap.manipulator): 72.06899999999999ms
{:name=>:sitemap_ignore}
== Instrument (sitemap.manipulator): 48.135ms
{:name=>:front_matter}
== Instrument (sitemap.manipulator): 91.626ms
{:name=>:sprockets}
== Instrument (sitemap.manipulator): 238.726ms
{:name=>:blog_portfolio_articles}
== Instrument (sitemap.manipulator): 43.452ms
{:name=>:blog_portfolio_paginate}
== Instrument (sitemap.update): 650.1139999999999ms
{:reasons=>[:registered_new_manipulator_i18n, :first_run_change_ignore, :registered_new_manipulator_blog_portfolio_articles, :registered_new_manipulator_blog_portfolio_tags, :registered_new_manipulator_blog_portfolio_calendar, :registered_new_manipulator_blog_portfolio_paginate]}

**BIGGEST PAUSE IS HERE**

== View your site at "http://Ians-Air.home:4567", "http://192.168.1.91:4567"
== Inspect your site configuration at "http://Ians-Air.home:4567/__middleman", "http://192.168.1.91:4567/__middleman"

**Comment by iwarner** [comment]
1st load of site

== Instrument (builder.output.resource.render-template): 1321.569ms
{:path=>"index.html.haml"}
== Instrument (builder.output.resource.render-layout): 555.947ms
{:path=>"fullWidth.haml"}

**Comment by iwarner** [comment]
Touched file

== Instrument (sitemap.manipulator): 42.531ms
{:name=>:sitemap_ondisk}
== Instrument (sitemap.manipulator): 85.467ms
{:name=>:sitemap_ignore}
== Instrument (sitemap.manipulator): 145.405ms
{:name=>:front_matter}
== Instrument (sitemap.manipulator): 101.158ms
{:name=>:sprockets}
== Instrument (sitemap.manipulator): 93.127ms
{:name=>:blog_portfolio_articles}
== Instrument (sitemap.manipulator): 54.661ms
{:name=>:blog_portfolio_paginate}
== Instrument (sitemap.update): 640.165ms
{:reasons=>[:touched_file]}

Refresh - This takes 12 seconds - but below numbers print out before loaded.

== Instrument (builder.output.resource.render-template): 1237.437ms
{:path=>"index.html.haml"}
== Instrument (builder.output.resource.render-layout): 545.586ms
{:path=>"fullWidth.haml"}

When no file is edited - get same as above really

== Instrument (builder.output.resource.render-template): 1275.722ms
{:path=>"index.html.haml"}
== Instrument (builder.output.resource.render-layout): 596.763ms
{:path=>"fullWidth.haml"}

**Comment by tdreyno** [comment]
But only after touching a file (source/index.html.haml in this case)

**Comment by iwarner** [comment]
Ahh I didn't see that in my output  - are you using my project ? Let me try again.

It would make sense that the assets are the issue here, but when editing a HAML should it rebuild the assets all the time?

**Comment by iwarner** [comment]
I see updating middleman:sprockets now - thanks steven for adding the timings. I hope this is helping guys, so important to get the speed up on middleman.

**Comment by iwarner** [comment]
== Instrument (sprockets.asset_lookup): 2526.198ms
{:asset=>#<Middleman::Sprockets::Resource path=assets/stylesheets/all.css>}
== Instrument (sprockets.render): 2526.366ms
{:asset=>#<Middleman::Sprockets::Resource path=assets/stylesheets/all.css>}
== Instrument (sprockets.asset_lookup): 7132.2660000000005ms
{:asset=>#<Middleman::Sprockets::Resource path=assets/javascripts/body.js>}
== Instrument (sprockets.render): 7132.451ms
{:asset=>#<Middleman::Sprockets::Resource path=assets/javascripts/body.js>}

**Comment by tdreyno** [comment]
Yeah, I don't use Sprockets, so this stuff flies pretty under my radar. It does seem like the haml file change is somehow clearing a cache on the sprockets side.

Sprockets has always been very slow, I wonder if they're doing something horrible like searching all of node_modules/bower_components for something when doing a lookup.

**Comment by stevenosloan** [comment]
Yeah sprockets getting slow after an html file is touched is my fault. As a quick fix for getting sass partials working I have the cache getting cleared on all file reloads. Working on making that a bit smarter, but so far am only able to improve on file changes that don't cause a server reload.

Speaking of, @tdreyno is there any way to persist something between application reloads? It'd be great to be able to keep the in-memory cache around. I've looked at using the file cache again -- but it was [https://github.com/middleman/middleman-sprockets/issues/53], and appears to still have some tough to tackle issues.

p.p.s
Those specific instrumentation numbers are a *little* misleading in this case since the asset lookup (which includes a render in Sprockets) is done inside the render method.

**Comment by iwarner** [comment]
@tdreyno I do also have sprockets.append_path File.join( root, "bower_components" ) so I guess I could amplify that problem also.

I will need to check out the new asset pipeline some more, but I pull .scss and .js files from all over the place so just didnt have the time to set it all up.

Thanks @stevenosloan let me know if there is anything else you want me to test @ me here.

**Comment by stevenosloan** [comment]
@iwarner if you don't mind could you open an issue on the middleman-sprockets project? I think it's fairly clear the issue is coming from there.

**Comment by iwarner** [comment]
@stevenosloan done

**Comment by tdreyno** [comment]
@stevenosloan Does a class accessor on your extension module persist between reloads? I never quite know what's happening in which thread, but I would think anywhere global would persist.

**Comment by stevenosloan** [comment]
Yeah, that totally works -- just need to give it some thought on properly invalidating that / keeping references to linked assets around. Wasn't sure if there were some built in lifecycle hooks to use that I wasn't aware of.

**Comment by dsignr** [comment]
Just in case anyone finds this useful - Middleman shutdown (on projects with sprockets) is quite slow too, sometimes even non-responsive, so I just use `killall -9 middleman` 🎯
